src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.cs

```diff
@@ -105,8 +104,8 @@ public override ImmutableArray<Location> Locations
                 if (_locations.IsDefault)
                 {
                     ImmutableInterlocked.InterlockedCompareExchange(ref _locations,
-                        _mergedDeclaration.NameLocations,
-                        default(ImmutableArray<Location>));
+                        IsInjected ? ImmutableArray.Create(Location.None) : _mergedDeclaration.NameLocations,
```

Comment:
>ImmutableArray.Create(Location.None) [](start = 37, length = 36)

I think in similar cases we return an empty array, rather than an array with a bogus location.

Reply:
I found that many places encode the assumption that a SourceNamespaceSymbol has at least one location.
For instance, we'll do `Locations[0]` without check when producing diagnostics.

---
In reply to: [215449429](https://github.com/dotnet/roslyn/pull/29180#discussion_r215449429) [](ancestors = 215449429)

Reply:
This doesn't mean that we should simply stick something into the array. This is just another argument for not using SourceNamespaceSymbol for injected namespaces. There is too much baggage. For example, some other consumer might not expect empty ```DeclaringSyntaxReferences```.

---
In reply to: [215460694](https://github.com/dotnet/roslyn/pull/29180#discussion_r215460694) [](ancestors = 215460694,215449429)

Reply:
For `DeclaringSyntaxReferences`, I didn't find users that directly index into a namespace symbol without a check.

Just to clarify, what is the type hierarchy you're proposing:
1) SourceNamespaceSymbol becomes a base and we have two concrete implementations
2) SourceNamespaceSymbol is one of two concrete implementations

The reason I'm asking is we have many places that type-check for SourceNamespaceSymbol. Either way, we have to validate a number of assumptions.
The current design feels safer and I've already ironed out a number of kinks. Alternative designs open up more risk, but the benefits are not quite clear to me yet.

---
In reply to: [215461552](https://github.com/dotnet/roslyn/pull/29180#discussion_r215461552) [](ancestors = 215461552,215460694,215449429)

Reply:
There's around 200 places in the CSharpCodeAnalysis that just do `Locations[0]` or `Locations.First()`. I don't know how many of those might be produced on injected namespaces or types, I expect only some low number.

We can either:
1. leave one `Location.None` on injected namespaces and types
2. review all those places to see if they could be reached and fix them with `Locations.FirstOrDefault() ?? Location.None`
3. fix all those places regardless

---
In reply to: [215462816](https://github.com/dotnet/roslyn/pull/29180#discussion_r215462816) [](ancestors = 215462816,215461552,215460694,215449429)

Reply:
I'll stop by in the morning (likely at 11) to discuss the need for a new symbol for injected namespace. That's the change that worries me (in terms of time, uncertainty and churn).

Reply:
I'm not sure why, but I'm not hitting the crashes I had encountered with `Locations[0]`. 
I filed a follow-up issue to replace all `Locations[0]` with something more robust.

Filed [https://github.com/dotnet/roslyn/issues/29707](https://github.com/dotnet/roslyn/issues/29707)

---
In reply to: [215477102](https://github.com/dotnet/roslyn/pull/29180#discussion_r215477102) [](ancestors = 215477102,215462816,215461552,215460694,215449429)